### PR TITLE
Enh: free busy UI

### LINKED ISSUE #39: Disable buttons after clicking to prevent double click actions
- [ ] calendar editor
- [ ] events editor popover
- [ ] events editor sidebar

### PR DESCRIPTION
**Pull request by hamza221**
Fixes #39
| Current state | Goal |
|--------|--------|
|https://github.com/nextcloud/calendar/assets/40746210/50a499d4-4c40-413e-b171-a8d76843c762|image| 

Todo list from design team: 
--***Nimisha***
Differences from current free/busy view:
- [x] Use components for all the elements (eg, date picker, today/previous day/next day buttons)
- [x] Add a heading "Find a time"
- [x] Add the time and time zone below the date in the top left corner
- [x] Nice-to-haves:
  - [x] The "guide" for the color (currently below the free/busy section) can be shifted into a small help section on the top right, which on hover opens a tooltip showing the guide 
  - [x] change the colors: 
    - [x] busy: `--color-error` in 0.3 opacity
    - [x] out of office/unavailable: `--color-background-darker`
    - [x] busy tentatively: `--color-warning` in 0.3 opacity
    - [x] unknown: `--color-info` in 0.3 opacity

New features:
- [x] Allow selecting different times in the modal itself: the selected time is indicated by the border (2px solid --color-primary-element with --border-radius)
- [x] Selected time is also reflected in the top left date and time
- [x] Show detailed attendees: Attendee names are shown as well as avatar and role, similar to how it's shown in the attendee list in the sidebar 
- [x] Add and remove attendees: Add attendees via a select component, similar to the attendee list in the sidebar, 
- [x] newly added attendees are shown at the bottom
- [x] Remove attendees using a close button
- [x] A done button to save changes 
-- ***Jan***
- [x] Scroll point: Currently when opening the free/busy modal, it always starts at midnight. Instead, it should scroll so the current time slot is in the middle. If it’s a full-day event, it could start at e.g. 8/9/10 AM, something of the sort. (If the person has out of office set, at their start of day for example)
- [ ] Color contrast: We need to make sure that the out of office, busy etc. blocks have proper contrast and/or are understandable without color. E.g. for out of office we could use grey with diagonal hatch pattern lines, and for busy we could use the red with diagonal criss-cross lines.
- [ ] Width: Since this element will most likely always scroll horizontally, it would be good to give it more space to the sides. Currently it is a box inside the modal, but it could also extend to the left and right edge (the text on the left should still be aligned with the other text of course). This is mostly a detail though, and least important of the points. :)

### CODE REVIEW

**Comment by tcitworld** [comment]
> Adding the delete option to fullCalendar is tricky since it's a premium feature 

We are entitled to the premium features: https://fullcalendar.io/license

https://github.com/nextcloud/calendar/blob/3bc0d2d59a0c3da04140655155fd963aaf27b5ef/src/components/Editor/FreeBusy/FreeBusy.vue#L223

**Comment by ChristophWurst** [comment]
- [ ] Nitpick: remove ":" at Bildschirmfoto vom 2024-02-27 11-56-08
- [ ] Bug: text is just cut and not ellipsized at Bildschirmfoto vom 2024-02-27 11-56-02
- [ ] Bug: only have border radius for top left and top right of first and bottom left and bottom right for last row of the fake event at Bildschirmfoto vom 2024-02-27 11-55-52

File: css/freebusy.scss [hunk 1]
```diff
@@ -28,7 +28,8 @@
 	}
 
 	.blocking-event-free-busy {
-		border-color: red;
+		border-radius: var(--border-radius);
+		border-color: var(--color-primary-element);
```
Comment (ChristophWurst) [inline css/freebusy.scss:32]: Bildschirmfoto vom 2024-02-27 11-55-52

fine-tune this for the --first-row and --last-row only

File: src/components/Editor/FreeBusy/FreeBusy.vue [hunk 2]
```diff
@@ -28,32 +28,104 @@
...
+					<NcButton type="secondary"
+						@click="handleActions('right')">
+						<template #icon>
+							<ChevronRightIcon :size="20" />
+						</template>
+					</NcButton>
+
+					<NcDateTimePicker :value="currentDate"
+						confirm
+						@confirm="(date)=>handleActions('picker',date)" />
```
Comment (ChristophWurst) [inline src/components/Editor/FreeBusy/FreeBusy.vue:75]:
```suggestion
						@confirm="(date)=>handleActions('picker', date)" />
```

**Review by ChristophWurst — CHANGES_REQUESTED** [review]

**Comment by hamza221** [comment]
> Bad contrast of calendar icon

It's a bug in nc/vue

File: src/components/Editor/FreeBusy/FreeBusy.vue [hunk 3]
```diff
@@ -254,39 +380,100 @@ export default {
...
+			}
+			&__timezone{
+				color: var(--color-text-lighter);
+			}
+		}
+	}
+}
+:deep(.vs__search ) {
+	text-overflow: ellipsis;
 }
```
Comment (ChristophWurst) [inline src/components/Editor/FreeBusy/FreeBusy.vue:476]: is this a generic issue with NcSelect? then let's make sure this is tracked and fixed upstream

we can keep this as temporary fix

**Comment by ChristophWurst** [comment]
> Scroll point

Is it fixable? That was the highest priority change requested by Jan. We should have this here or in a follow-up.

File: src/components/Editor/FreeBusy/FreeBusy.vue [hunk 4]
```diff
@@ -28,32 +28,104 @@
 			<div v-if="loadingIndicator" class="loading-indicator">
 				<div class="icon-loading" />
 			</div>
+			<div class="modal__content__header">
+				<h2>{{ t('calendar','Find a time') }}</h2>
```
Comment (ChristophWurst) [inline src/components/Editor/FreeBusy/FreeBusy.vue:32]:
```suggestion
				<h2>{{ t('calendar', 'Find a time') }}</h2>
```

we need better eslint config

File: src/components/Editor/FreeBusy/FreeBusy.vue [hunk 5]
```diff
@@ -28,32 +28,104 @@
...
+				<h3>{{ eventTitle }}</h3>
+				<div class="modal__content__header__attendees">
+					{{ t('calendar','with') }}
```
Comment (ChristophWurst) [inline src/components/Editor/FreeBusy/FreeBusy.vue:35]:
```suggestion
					{{ t('calendar', 'with') }}
```

**Review by ChristophWurst — APPROVED** [review]
Works :rocket: 

Bildschirmfoto vom 2024-02-28 13-23-08

^ IMO needs a bit of padding but we can leave that open until the design review

Great work!

File: src/components/Editor/FreeBusy/FreeBusy.vue (hunk 3 above)
Comment (hamza221) [reply src/components/Editor/FreeBusy/FreeBusy.vue:486]: https://github.com/nextcloud-libraries/nextcloud-vue/pull/5325

**Pull request merged**